examples/vortex-mpi.py

```diff
@@ -147,56 +151,57 @@ def main(ctx_factory=cl.create_some_context, use_profiling=False, use_logmgr=Fal
         vis_timer = IntervalTimer("t_vis", "Time spent visualizing")
         logmgr.add_quantity(vis_timer)
 
-    visualizer = make_visualizer(discr, order + 3 if dim == 2 else order)
-
-    initname = initializer.__class__.__name__
-    eosname = eos.__class__.__name__
-    init_message = make_init_message(dim=dim, order=order,
+    init_message = make_init_message(dim=dim, order=order, casename=casename,
                                      nelements=local_nelements,
-                                     global_nelements=global_nelements,
-                                     dt=current_dt, t_final=t_final, nstatus=nstatus,
-                                     nviz=nviz, cfl=current_cfl,
-                                     constant_cfl=constant_cfl, initname=initname,
-                                     eosname=eosname, casename=casename)
+                                     global_nelements=global_nelements)
     if rank == 0:
         logger.info(init_message)
 
-    get_timestep = partial(inviscid_sim_timestep, discr=discr, t=current_t,
-                           dt=current_dt, cfl=current_cfl, eos=eos,
-                           t_final=t_final, constant_cfl=constant_cfl)
+    visualizer = make_visualizer(discr, order + 3 if dim == 2 else order)
+
+    def write_vis(step, t, state):
+        io_fields = get_inviscid_vis_fields(dim, state, eos)
+        io_fields.append(
+            ("exact_cv", split_conserved(dim, initializer(nodes, t=t))))
+        return write_visualization_file(visualizer, fields=io_fields,
+                    basename=casename, step=step, t=t, comm=comm, timer=vis_timer)
 
-    def my_rhs(t, state):
+    def get_timestep(step, t, state):
+        try:
+            dt_max = get_inviscid_timestep(discr=discr, q=state, cfl=current_cfl,
+                eos=eos) if constant_cfl else current_dt
+        except InviscidTimestepError:
+            write_vis(step, t, state)
+            raise
+        return get_sim_timestep(dt_max, t, t_final=t_final)
```

Comment:
This seems like it leaves too much for the user to get right in order to have timestepping work correctly.  That is my only real concern about this.